pkg/scheduler/core/division_algorithm.go

```diff
@@ -20,6 +20,82 @@ func (a TargetClustersList) Len() int           { return len(a) }
 func (a TargetClustersList) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a TargetClustersList) Less(i, j int) bool { return a[i].Replicas > a[j].Replicas }
 
+type dispenser struct {
+	numReplicas int32
+	result      []workv1alpha2.TargetCluster
+}
+
+func newDispenser(numReplicas int32, init []workv1alpha2.TargetCluster) *dispenser {
+	cp := make([]workv1alpha2.TargetCluster, len(init))
+	copy(cp, init)
+	return &dispenser{numReplicas: numReplicas, result: cp}
+}
+
+func (a *dispenser) done() bool {
+	return a.numReplicas == 0 && len(a.result) != 0
+}
+
+func (a *dispenser) takeByWeight(w helper.ClusterWeightInfoList) {
+	if a.done() {
+		return
+	}
+	sum := w.GetWeightSum()
+	if sum == 0 {
+		return
+	}
+
+	sort.Sort(w)
+
+	result := make([]workv1alpha2.TargetCluster, 0, w.Len())
+	remain := a.numReplicas
+	for _, info := range w {
+		replicas := int32(info.Weight * int64(a.numReplicas) / sum)
+		result = append(result, workv1alpha2.TargetCluster{
+			Name:     info.ClusterName,
+			Replicas: replicas,
+		})
+		remain -= replicas
+	}
+	// TODO(Garrybest): take rest replicas by fraction part
+	for i := range result {
+		if remain == 0 {
+			break
+		}
+		result[i].Replicas++
+		remain--
+	}
+
+	a.numReplicas = remain
```

Comment:
What's the function of step `a.numReplicas = remain` and `if a.done() {`? In waht kind of scenario, it will be used

Reply:
It's a reservation here. The `dispenser` is designed for dispensing in multiple times considering constraints in the future. If we have some contraints, one-time dispensing may not finish yet. So we need this function here to conclude whether the dispensing is done.

Reply:
I prefer to delete it, if we need it, then we add it back.

Reply:
All right.

Reply:
Hi @jwcesign, I have reconsidered this again. From the aspect of a go `struct`, if the invoker invokes function `takeByWeight` repeatedly, the result may be a mess because the dispensing action is done before unless invoker set a new `numReplicas`. So I think in terms of object-oriented concept, this function `done()` is vital and we never do it twice if `numReplicas` has been dispensed. How do you think? 😊

Reply:
What I'm thinking about is how to use this function. I see that every time this function is called, a new structure will be created, so there is no concurrency problem(or assign two times).

Reply:
You are right, from the viewpoint of invoker. Nevertheless, for the design of this struct, it does not know how the invoker will use the struct. So `done` is just an assurance for check here, I still consider it as a reasonable desgin here. 🤣